docs/datatypes/contextual/index.md

```diff
@@ -323,24 +225,102 @@ object LoggingLive {
 }
 ```
 
-5. **Accessor Methods** — Finally, to create the API more ergonomic, it's better to write accessor methods for all of our service methods. Just like what we did in Module Pattern 1.0, but with a slight change, in this case, instead of using `ZIO.accessZIO` we use `ZIO.serviceWithZIO` method to define accessors inside the service companion object:
+5. **Accessor Methods** — Finally, to create the API more ergonomic, it's better to write accessor methods for all of our service methods using `ZIO.serviceWithZIO` constructor inside the companion object:
 
 ```scala mdoc:silent
 object Logging {
   def log(line: String): URIO[Logging, Unit] = ZIO.serviceWithZIO[Logging](_.log(line))
 }
 ```
 
+Accessor methods allow us to utilize all the features inside the service through the ZIO Environment. That means, if we call `Logging.log`, we don't need to pull out the `log` function from the ZIO Environment. The `ZIO.serviceWithZIO` constructor helps us to access the environment and reduce the redundant operations, every time.
+
+This is how ZIO services are created. Let's use the `Logging` service in our application. We should provide the live layer of `Logging` service to be able to run the application:
+
+```scala mdoc:compile-only
+import zio._
+import java.io.IOException
+
+object MainApp extends ZIOAppDefault {
+  val app: ZIO[Logging with Console, IOException, Unit] =
+    for {
+      _    <- Logging.log("Application Started!")
+      _    <- Console.print("Enter your name:")
+      name <- Console.readLine
+      _    <- Console.printLine(s"Hello, $name!")
+      _    <- Logging.log("Application Exited!")
+    } yield ()
+
+  def run = app.provideCustom(LoggingLive.layer)
+}
+```
+
+During writing the application, we don't care which implementation version of the `Logging` service will be injected into our `app`, later at the end of the day, it will be provided by one of `ZIO#provide*` methods.
+
 That's it! Very simple! ZIO encourages us to follow some of the best practices in object-oriented programming. So it doesn't require us to throw away all our object-oriented knowledge.
 
-Finally, we provide required layers to our `app` effect:
+## The Three Laws of ZIO Environment
 
-```scala
- val app = Logging.log("Application Started")
+When we are working with the ZIO environment, one question might arise: "When should we use environment and when do we need to use constructors?".
+
+Using ZIO environment follows three laws:
+
+1. **Service Interface (Trait)** — When we are defining service interfaces, we shouldn't care about implementation details like what dependencies we have. So at this level, we should _never_ use the environment. For example, the following service definition is wrong:
```

Comment:
Let's clarify a bit here that we should _never_ use the environment for *dependencies of the service itself*.

For example, if implementation of service X depends on services Y and Z, then these should *never* be reflected in the trait that defines Service X. It's leaking implementation details.

In rare occasions, all of which involve _local context_ that is _independent of implementation_, it's acceptable to use the environment in the definition of a service.

Two examples come to mind:

1. In a web application, a service may be defined only to operate in the context of an HTTP request. In such a case, the request itself could be stored in the environment: `ZIO[HttpRequest, ...]`. This is acceptable because this use of the environment is part of the semantics of the trait itself, rather than leaking an implementation detail of some particular class that implements the service trait.
2. In a database application, a service may be defined only to operate in the context of a larger database transaction. In such a case, the transaction could be stored in the environment: `ZIO[DatabaseTransaction, ...]`. As in the previous example, because this is part of the semantics of the trait itself (whose functionality all operates within a transaction), this is not leaking implementation details and therefore it is valid.

While it's better to err on the side of "don't put it into the environment", there are cases where it's acceptable, and we should provide a few examples so more expert users have guidance on best practices.

Reply:
Thank you for your explanations. I addressed them with two examples.